### PR TITLE
Describe our blockchain structure in detail.

### PR DESCRIPTION
**Pull request by chb0github**
## Description

Fixes pyrsia/pyrsia#

This PR does... by accomplishing ... and it can be reviewed by ...

## PR Checklist

- [x] I've read the [contributing guidelines](https://github.com/pyrsia/.github/blob/main/contributing.md).
- [x] I've read ["What is a Good PR?"](https://github.com/pyrsia/pyrsia/blob/main/docs/good_pr.md)
- [x] I've included a good title and brief description along with how to review them.
- [x] I've linked any associated an [issue](https://github.com/pyrsia/pyrsia/issues).
- [x] I've requested a review  from "pyrsia/collaborators"

### CODE REVIEW

File: docs/blockchain/whats_in_the_blockchain.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+we found as inspired. Where possible, we tried to dig into reasons for any particular choice. But ultimately we decided that what we were seeing represented
+state-of-the-art, and we should lean towards accepting it.
+
+We looked at the source code for `ETH` and several minor OSS implementations.
+
+# Prior art
+
+When constructing this blockchain we consulted the structure of several famous blockchain implementations. Namely
+`BTC` and `ETH`. Bitcoin was the simplest, but it has the worst growing pains and doesn't handle chain merging well.
+Etherium manages chain merging by treating it not as a blockchain, but as a block tree - `Merkle tree`. In addition, all of the
```
Comment (fishseabowl) [inline docs/blockchain/whats_in_the_blockchain.md:28]: Please double check. I think Bitcoin also support Merkle tree.

Comment (chb0github) [reply docs/blockchain/whats_in_the_blockchain.md:28]: Ok, I am verifying.

File: docs/blockchain/whats_in_the_blockchain.md [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+new implementations of blockchain allow for a list of transactions in a single block. This keep the block count lower and thus
+increases overall efficiency since it's the blocks that require consensus and take time.
+
+TO that end, our blockchain contains:
```
Comment (johanvos) [inline docs/blockchain/whats_in_the_blockchain.md:32]:
```suggestion
To that end, our blockchain contains:
```

File: docs/blockchain/whats_in_the_blockchain.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+2. Signed transaction hash for efficiently verifying transactions
+3. Merkle root indicator so that we can use a merkle tree implementation
+
+# Unresolved questions
+
+None.
+
+# Future possibilities
+
+This structure is most definitely going to change as we learn and grow. There are already several outstanding tickets to do so
```
Comment (johanvos) [inline docs/blockchain/whats_in_the_blockchain.md:43]: Maybe add links to those tickets?

File: docs/blockchain/whats_in_the_blockchain.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+Our blockchain consists
+
+* series of blocks where
+    * each block has
+        * a Header
+            * which contains it's parent hash. This is the part of the block chain that makes it a `chain`. Each block has a reference to it's parent going all the way back to the genesis. Basically an inductive proof.
+            * The block committers address. On any block chain you need to identify the participant. The address isn't really any kind of routing destination or origin. It's the unique id of a participant and nothing more.
+            * ~the transactions root - This is an optimization to verify all transactions in a block. We can simply add the transaction hash (256bit Keccak Hash of the root node of Transactions Merkle tree). Certain block chain implementations aren't blocks, they are actually merkle trees (this is done to handle chain merging conflicts)~
+            * the timestamp of, in epoch time in seconds, of when the block was created
+            * a header number is a monotonically increasing ordinal. Used to sort the blocks
```
Comment (johanvos) [inline docs/blockchain/whats_in_the_blockchain.md:56]: Is this guaranteed to be unique?

File: docs/blockchain/whats_in_the_blockchain.md [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+            * a `nonce` for salting the payload
+            * a Block header hash - a singular unique hash of all the data in the block header (except, of course, the signature)
+        * a series of transactions
+            * Each transaction has a type (`AddAuthority`, `RevokeAuthority`, `Generic`) - that help determine how the payload should be interpreted.
+            * The submitter of the transaction, represented as a blockchain address
+            * The timestamp of the transaction in epoch time
+            * the actual payload as a series of bytes
+            * a `nonce` used to salt the signature of the block
```
Comment (johanvos) [inline docs/blockchain/whats_in_the_blockchain.md:64]: There is a nonce for the block + a nonce for every transaction, correct?

File: docs/blockchain/whats_in_the_blockchain.md [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+- RFC PR: [pyrsia/pyrsia#0000](https://github.com/pyrsia/pyrsia/pull/0000)
+- Pyrsia Issue: [pyrsia/pyrsia#0000](https://github.com/pyrsia/pyrsia/issues/0000)
+
+# Summary
+
+What's in the Pyrsia blockchain and why
+
+# Motivation
+
+In order to make a robust blockchain, and give the community some guidance on our decisions, documenting each field
```
[the remark this answers is not in the document]
Comment (prince-chrismc) [reply docs/blockchain/whats_in_the_blockchain.md:12]: The intention is to not limit the data that can be stored. There will be a "generic" payload that can be recorded.

> what information Pyrsia plans to store

I think needs to be answered by the metadata/docker data not by the blockchain (which does not know about the Pyrsia application layer sitting on top)

File: docs/blockchain/whats_in_the_blockchain.md (hunk 4 above)
Comment (prince-chrismc) [reply docs/blockchain/whats_in_the_blockchain.md:56]: Yes! but I think this is coming out as a requirement here.

TL;DR is using Aleph we get a pure BFT, where every block is sequentially ordered by the consensus. We only do one block at a time.

File: docs/blockchain/whats_in_the_blockchain.md (hunk 5 above)
Comment (prince-chrismc) [reply docs/blockchain/whats_in_the_blockchain.md:64]: Yes! 

- Transactions can handled independently so they get one 
https://github.com/prince-chrismc/pyrsia-sandbox/blob/844d1f1c4687afabec320584cea6d57b218fe1c4/src/blockchain/src/structures/transaction.rs#L88

- The blocks header also has a nonce
https://github.com/prince-chrismc/pyrsia-sandbox/blob/844d1f1c4687afabec320584cea6d57b218fe1c4/src/blockchain/src/structures/header.rs#L75

***

Both are published by participants and signed, and they can be seen by others so they'll get the same "protection" features

File: docs/blockchain/whats_in_the_blockchain.md [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+## A structural breakdown of our blockchain ##
+
+Our blockchain consists
+
+* series of blocks where
+    * each block has
+        * a Header
+            * which contains it's parent hash. This is the part of the block chain that makes it a `chain`. Each block has a reference to it's parent going all the way back to the genesis. Basically an inductive proof.
+            * The block committers address. On any block chain you need to identify the participant. The address isn't really any kind of routing destination or origin. It's the unique id of a participant and nothing more.
+            * ~the transactions root - This is an optimization to verify all transactions in a block. We can simply add the transaction hash (256bit Keccak Hash of the root node of Transactions Merkle tree). Certain block chain implementations aren't blocks, they are actually merkle trees (this is done to handle chain merging conflicts)~
```
Comment (prince-chrismc) [inline docs/blockchain/whats_in_the_blockchain.md:54]: This is being used still post #517

File: docs/blockchain/whats_in_the_blockchain.md [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+This structure is most definitely going to change as we learn and grow. There are already several outstanding tickets to do so
+
+## A structural breakdown of our blockchain ##
+
+Our blockchain consists
+
+* series of blocks where
+    * each block has
+        * a Header
```
Comment (prince-chrismc) [inline docs/blockchain/whats_in_the_blockchain.md:51]: It's worth mentioning the headers are what can be used to do a quick verification of the chain

File: src/blockchain/src/blockchain.rs [hunk 9]
```diff
@@ -103,14 +109,15 @@ impl Blockchain {
                     PartialTransaction::new(
                         TransactionType::AddAuthority,
                         local_id,
-                        "this needs to be the root authority".as_bytes().to_vec(),
+                        genesis_pub_key.to_vec(),
```
Comment (prince-chrismc) [inline src/blockchain/src/blockchain.rs:112]: You pushed changes from your other branch 😕

**Review by johanvos — DISMISSED** [review]

File: src/blockchain/src/blockchain.rs (hunk 9 above)
Comment (chb0github) [reply src/blockchain/src/blockchain.rs:112]: I don't see how. Infuriating. Let me see if I can fix this

File: docs/blockchain/whats_in_the_blockchain.md (hunk 4 above)
Comment (chb0github) [reply docs/blockchain/whats_in_the_blockchain.md:56]: Well, it was/is a requirement for Aleph. So, now I am stating it as fact

File: docs/blockchain/whats_in_the_blockchain.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (chb0github) [reply docs/blockchain/whats_in_the_blockchain.md:12]: I will add some wording here to make this explicit

File: docs/blockchain/whats_in_the_blockchain.md (hunk 3 above)
Comment (chb0github) [reply docs/blockchain/whats_in_the_blockchain.md:43]: I will nix the words about outstanding.. this list is going to disappear quickly and thus leave nothing to refer to
